### ISSUE TITLE
VideoCapture FFmpeg backend color shift when reading H.264 video with BT.709 color standard

### ISSUE THREAD

**Issue by cohenrotem** [opencv/opencv#20513]
##### System information (version)
- OpenCV => 4.53
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2017

##### Detailed description

There is an issue with OpenCV VideoCapture when reading video frames using FFmpeg backend.  
The issue results a "color shift" when reading H.264 video stream that marked as [BT.709](https://en.wikipedia.org/wiki/Rec._709) color standard.  

The problem is reproduced (at least) under Windows 10 in C++ and Python interfaces and under Ubuntu 18.04 with Python interface.  

The issue is described in my answer in following post: [OpenCV read / write video color difference](https://stackoverflow.com/questions/68629271/opencv-read-write-video-color-difference).  

It looks like OpenCV video frame capturing ignores the color characteristics of the H.264 video stream, and applies YUV to RGB color conversion using BT.601 color conversion formula.  
Color description in a stream is described in H.264 section E.2.1 and tables E-3, E-4 and E-5.

##### Steps to reproduce

**Building synthetic video pattern** for reproducing the problem:
Use FFmpeg command line tool for creating a synthetic video to be used as input.
The following command generates an MP4 video file with H.264 codec, and BT.709 color standard:

`ffmpeg -y -f lavfi -src_range 1 -color_primaries bt709 -color_trc bt709 -colorspace bt709 -i testsrc=size=192x108:rate=1:duration=5 -vcodec libx264 -crf 17 -pix_fmt yuv444p -dst_range 1 -color_primaries bt709 -color_trc bt709 -colorspace bt709 -bsf:v h264_metadata=video_full_range_flag=1:colour_primaries=1:transfer_characteristics=1:matrix_coefficients=1 bt709_full_range.mp4`

- The above command uses yuv444p pixel format (instead of yuv420p) for getting more pure colors.
- The arguments -bsf:v h264_metadata=video_full_range_flag=1:colour_primaries=1:transfer_characteristics=1:matrix_coefficients=1 use Bitstream Filter for marking the H.264 stream as "full range" BT.709.

Using [MediaInfo](https://mediaarea.net/en/MediaInfo) tool, we can view the following color characteristics:

`colour_range:             Full 
colour_primaries:         BT.709 
transfer_characteristics: BT.709 
matrix_coefficients:      BT.709 
`

**Capturing the video using OpenCV:**  

The following C++ code grabs the first frame, and save it to `1.png` image file:

#include "opencv2/opencv.hpp"

```c++
void main()
{   
    cv::VideoCapture cap("bt709_full_range.mp4");

    cv::Mat frame;    
    cap >> frame;

    cv::imwrite("1.png", frame);

    cap.release();    
}
```

Converting bt709_full_range.mp4 into images sequence using FFmpeg:

`ffmpeg -i bt709_full_range.mp4 -pix_fmt rgb24 %03d.png`

The file name of the first "extracted" frame is 001.png.

 - The value of the red color pixels of OpenCV are RGB = [232, 0, 3].
 - The value of the red color pixels of FFmpeg are RGB = [254, 0, 0].
The original RGB value is probably [255, 0, 0] (value is 254 due to colors conversion).

Comparing the results:

 - The left side is 1.png (result of OpenCV)
 - The right side is 001.png (result of FFmpeg command line tool)

![1](https://user-images.githubusercontent.com/20123928/128600632-c8f20cb7-ecea-4ffb-bb56-470444825a3e.png)  ![2](https://user-images.githubusercontent.com/20123928/128600644-52eaf82a-94e7-4155-87d9-47a2b1c31875.png)

Using GStreamer backend (instead of FFmpeg) backend seems to solve the "color shifting" problem.

**Comment by ValZapod** [comment]
> -bsf:v h264_metadata=video_full_range_flag=1:colour_primaries=1:transfer_characteristics=1:matrix_coefficients=1

Do not do this insanity please. Just use -color_range pc. Also that just tags the video, becuse -vf scale defaults to BT.601 limited range you need to use -vf scale=out_color_matrix=bt709:out_range=pc

Oogh. Seriously?

**Comment by cohenrotem** [comment]
Seriously?  

Tagging the video is the whole point.  
The main issue is that `VideoCapture` ignores the color tags, and that is the reason for the "color shift".  

The FFmpeg command line looks "insane", but it's not related to OpenCV issue, because the output is 100% valid video file.

The `-color_range` is not the issue here.  
It look like `VideoCapture` respects the `video_full_range_flag=1` tag, but ignores the `BT.709` color related tags.

**Comment by ValZapod** [comment]
> Tagging the video is the whole point.

No, it is not. Did you try to add -vf scale=out_color_matrix=bt709:out_range=pc?

**Comment by cohenrotem** [comment]
No... from the perspective of the issue the color tags are the point.  
Are you saying that the MP4 file is not valid?  
The MP4 file is valid - the video players plays it correctly, and GStreamer backend handles it correctly.  

I tried using the following command:  

`ffmpeg -y -f lavfi -i testsrc=size=192x108:rate=1:duration=5 -vf scale=out_color_matrix=bt709:out_range=pc -vcodec libx264 -crf 17 -pix_fmt yuv444p -bsf:v h264_metadata=video_full_range_flag=1:colour_primaries=1:transfer_characteristics=1:matrix_coefficients=1 bt709_full_range.mp4`  

The result is the same (the RGB value of the red pixels in OpenCV is [232, 0, 3]).

**Comment by ValZapod** [comment]
Correct command is ffmpeg -y -f lavfi -i testsrc=size=192x108:rate=1:duration=5 -vf scale=out_color_matrix=bt709:out_range=pc -vcodec libx264 -crf 17 -pix_fmt yuv444p  -color_primaries bt709 -color_trc bt709 -colorspace bt709 -color_range pc bt709_full_range.mp4
https://user-images.githubusercontent.com/31514790/129698344-cb97eaec-ae66-4490-afbb-0fccdec236bb.mp4


You first command as can be checked with ffplay bt709_full_range.mp4 -vf extractplanes=y produces limited range (see black is 16). And while comparing with https://www.itu.int/dms_pubrec/itu-r/rec/bt/R-REC-BT.1729-0-200504-I!!PDF-E.pdf (page 16, Table  5) your second command you can see it is BT.709 (I was wrong here, sigh) (indeed red color Y for BT.601 is 81, not 63).
And it is tagged as Color range: Limited.


Your second command produces correct stuff but it is not tagged on container.

My command produces perfect file with nclx atom.

**Comment by cohenrotem** [comment]
The relevant tags are embedded in the H.264 bitstream (not in the MP4 container).  
In case it bothers you, you may reproduce the problem with raw H.264 stream  (without container).  

`ffmpeg -y -f lavfi -i testsrc=size=192x108:rate=1:duration=5 -vf scale=out_color_matrix=bt709:out_range=pc -vcodec libx264 -crf 17 -pix_fmt yuv444p -bsf:v h264_metadata=video_full_range_flag=1:colour_primaries=1:transfer_characteristics=1:matrix_coefficients=1 bt709_full_range.264`

**Comment by ValZapod** [comment]
> not in the MP4 container

I know how H.264 stream works. I did look into it with trace_headers. The fact of the matter your first command produces absolutelly wrong file your second is not good enoigh.

**Comment by ValZapod** [comment]
I just reported another bug in ffmpeg that was tagging full range RGB planes as limited for x11grab to x264rgb. So as you can see there are a lot of bugs there and you need to be careful. As you can see in commit message it dirrectly affects VUI, i.e. what h264_metadata bsf does. https://github.com/FFmpeg/FFmpeg/commit/7ca71b79f2b3256a0eef1a099b857ac9e4017e36

**Comment by mirh** [comment]
Conversation did evolve [here](https://github.com/k4yt3x/video2x/issues/403#issuecomment-1012683319), and it's being tracked [here](https://trac.ffmpeg.org/ticket/9167).

**Comment by czk32611** [comment]
It has been 2 years. Do anyone know how to solve this problem?

**Comment by mirh** [comment]
The problem in ffmpeg has presumably been a thing for the better part of its very existence.. 
Anyway, I suppose you can bump the issue yourself if you are so pressed (I feel like I already did that myself too many times by now).

**Comment by Pamimi** [comment]
Is there a solution without building OpenCV by Gstream？
